features/lorawan/LoRaWANStack.cpp

```diff
@@ -122,56 +122,72 @@ void LoRaWANStack::bind_radio_driver(LoRaRadio& radio)
-    const static uint32_t nwk_id = (MBED_CONF_LORA_DEVICE_ADDRESS & LORAWAN_NETWORK_ID_MASK);
-
-    connection_params.connect_type = LORAWAN_CONNECTION_ABP;
-    connection_params.connection_u.abp.nwk_id = nwk_id;
-    connection_params.connection_u.abp.dev_addr = dev_addr;
-    connection_params.connection_u.abp.nwk_skey = const_cast<uint8_t *>(nwk_skey);
-    connection_params.connection_u.abp.app_skey = const_cast<uint8_t *>(app_skey);
-
-    return activation_by_personalization(connection_params);
-#endif
+    if (DEVICE_STATE_NOT_INITIALIZED == _device_current_state) {
+        tr_error("Stack not initialized!");
+        return LORAWAN_STATUS_NOT_INITIALIZED;
+    }
+
+    lorawan_status_t status = _loramac.prepare_join(NULL, MBED_CONF_LORA_OVER_THE_AIR_ACTIVATION);
+
+    if (LORAWAN_STATUS_OK != status) {
+        return status;
+    }
+
+    return handle_connect(MBED_CONF_LORA_OVER_THE_AIR_ACTIVATION);
 }
 
 lorawan_status_t LoRaWANStack::connect(const lorawan_connect_t &connect)
 {
-    lorawan_status_t mac_status;
+    if (DEVICE_STATE_NOT_INITIALIZED == _device_current_state) {
+        tr_error("Stack not initialized!");
+        return LORAWAN_STATUS_NOT_INITIALIZED;
+    }
+
+    if (!(connect.connect_type == LORAWAN_CONNECTION_OTAA) &&
+        !(connect.connect_type == LORAWAN_CONNECTION_ABP)) {
+        return LORAWAN_STATUS_PARAMETER_INVALID;
+    }
+    bool is_otaa = (connect.connect_type == LORAWAN_CONNECTION_OTAA);
+
+    lorawan_status_t status = _loramac.prepare_join(&connect, is_otaa);
+
+    if (LORAWAN_STATUS_OK != status) {
+        return status;
+    }
+
+    return handle_connect(is_otaa);
+}
+
+lorawan_status_t LoRaWANStack::handle_connect(bool is_otaa)
+{
+    device_states_t new_state;
+
+    if (is_otaa) {
+        tr_debug("Initiating OTAA");
 
-    if (connect.connect_type == LORAWAN_CONNECTION_OTAA) {
-        mac_status = join_request_by_otaa(connect);
-    } else if (connect.connect_type == LORAWAN_CONNECTION_ABP) {
-        mac_status = activation_by_personalization(connect);
+        // As mentioned in the comment above, in 1.0.2 spec, counters are always set
+        // to zero for new connection. This section is common for both normal and
+        // connection restore at this moment. Will change in future with 1.1 support.
+        _lw_session.downlink_counter = 0;
+        _lw_session.uplink_counter = 0;
+        new_state = DEVICE_STATE_JOINING;
     } else {
-        return LORAWAN_STATUS_PARAMETER_INVALID;
+        // If current state is SHUTDOWN, device may be trying to re-establish
+        // communication. In case of ABP specification is meddled about frame counters.
+        // It says to reset counters to zero but there is no mechanism to tell the
+        // network server that the device was disconnected or restarted.
+        // At the moment, this implementation does not support a non-volatile
+        // memory storage.
+        //_lw_session.downlink_counter; //Get from NVM
```

Comment:
why are these 2 lines here?

Reply:
This is a reminder for NVM support. It will be added later for LoRa